### PR TITLE
Fix the config for `jest` with `swc`

### LINKED ISSUE #13502: Jest SWC transpiling issues when mixing TSX and JSX in v1.6.0-next.1
## Expected Behaviour

Jest using SWC should be able to transpile a JSX file that is imported into a TSX file

## Actual Behaviour

SWC throws and error with "Syntax error" when it attempts to transpile JSX from within a TSX file

<details>
  <summary>Example test error output</summary>

```text
Test suite failed to run


      x Expression expected
       ,-[/home/runner/work/backstage-example/backstage-example/packages/app/src/components/HelloWorld.jsx:4:3]
     4 | return <h1>Hello World!</h1>
       :        ^
       `----

      x Expected ';', got 'World'
       ,-[/home/runner/work/backstage-example/backstage-example/packages/app/src/components/HelloWorld.jsx:4:3]
     4 | return <h1>Hello World!</h1>
       :                  ^^^^^
       `----


    Caused by:
        0: failed to process input file
        1: Syntax Error

      at Compiler.transformSync (../../../node_modules/@swc/core/index.js:241:29)
      at transformSync (../../../node_modules/@swc/core/index.js:34[8](https://github.com/UsainBloot/backstage-example/runs/8154029154?check_suite_focus=true#step:5:9):21)
      at Object.process (../../../node_modules/@swc/jest/index.js:72:45)
      at Object.process (../../../node_modules/@backstage/cli/config/jestSwcTransform.js:27:27)
      at ScriptTransformer.transformSource (../../../node_modules/@jest/transform/build/ScriptTransformer.js:61[9](https://github.com/UsainBloot/backstage-example/runs/8154029154?check_suite_focus=true#step:5:10):31)
      at ScriptTransformer._transformAndBuildScript (../../../node_modules/@jest/transform/build/ScriptTransformer.js:765:40)
      at ScriptTransformer.transform (../../../node_modules/@jest/transform/build/ScriptTransformer.js:822:[19](https://github.com/UsainBloot/backstage-example/runs/8154029154?check_suite_focus=true#step:5:20))
```
</details>

## Steps to Reproduce

See recreation https://github.com/UsainBloot/backstage-example/pull/3 with [GitHub Action output - test error](https://github.com/UsainBloot/backstage-example/runs/8154029154?check_suite_focus=true)

1. Update to 1.6.0-next.1
2. Add a JSX component to the App (https://github.com/UsainBloot/backstage-example/pull/3/commits/6bf1e68ec71b0599dabd326094c1bb4639c4a690)
3. Run tests ([Usainbloot/backstage-example - CI output](https://github.com/UsainBloot/backstage-example/runs/8154029154?check_suite_focus=true))

## Context

While attempting to upgrade my own instance of Backstage to use SWC, I observed errors when running tests that tested components that import a mixture of TSX and JSX.

**Note:** This is not an issue when importing a JS file into a TS/TSX file.

## Your Environment

<details>
  <summary>Output of `yarn backstage-cli info`:</summary>

```text
OS:   Darwin 21.6.0 - darwin/x64
node: v16.14.0
yarn: 1.22.19
cli:  0.19.0-next.1 (installed)
backstage:  1.6.0-next.1

Dependencies:
  @backstage/app-defaults                          1.0.6-next.1
  @backstage/backend-common                        0.15.1-next.1
  @backstage/backend-plugin-api                    0.1.2-next.0
  @backstage/backend-tasks                         0.3.5-next.0
  @backstage/catalog-client                        1.0.5-next.0
  @backstage/catalog-model                         1.1.0
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.19.0-next.1
  @backstage/config-loader                         1.1.4-next.0
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.1.0-next.1
  @backstage/core-components                       0.11.1-next.1
  @backstage/core-plugin-api                       1.0.6-next.1
  @backstage/errors                                1.1.0
  @backstage/integration-react                     1.1.4-next.0
  @backstage/integration                           1.3.1-next.0
  @backstage/plugin-api-docs                       0.8.9-next.1
  @backstage/plugin-app-backend                    0.3.36-next.1
  @backstage/plugin-auth-backend                   0.16.0-next.1
  @backstage/plugin-auth-node                      0.2.5-next.1
  @backstage/plugin-catalog-backend                1.4.0-next.1
  @backstage/plugin-catalog-common                 1.0.6-next.0
  @backstage/plugin-catalog-graph                  0.2.21-next.1
  @backstage/plugin-catalog-import                 0.8.12-next.1
  @backstage/plugin-catalog-node                   1.0.2-next.0
  @backstage/plugin-catalog-react                  1.1.4-next.1
  @backstage/plugin-catalog                        1.5.1-next.1
  @backstage/plugin-github-actions                 0.5.9-next.1
  @backstage/plugin-org                            0.5.9-next.1
  @backstage/plugin-permission-common              0.6.4-next.0
  @backstage/plugin-permission-node                0.6.5-next.1
  @backstage/plugin-permission-react               0.4.5-next.1
  @backstage/plugin-proxy-backend                  0.2.30-next.0
  @backstage/plugin-scaffolder-backend             1.6.0-next.1
  @backstage/plugin-scaffolder-common              1.2.0-next.0
  @backstage/plugin-scaffolder                     1.6.0-next.1
  @backstage/plugin-search-backend-module-pg       0.4.0-next.1
  @backstage/plugin-search-backend-node            1.0.2-next.1
  @backstage/plugin-search-backend                 1.0.2-next.0
  @backstage/plugin-search-common                  1.0.1-next.0
  @backstage/plugin-search-react                   1.0.2-next.1
  @backstage/plugin-search                         1.0.2-next.1
  @backstage/plugin-tech-radar                     0.5.16-next.1
  @backstage/plugin-techdocs-backend               1.2.2-next.0
  @backstage/plugin-techdocs-module-addons-contrib 1.0.4-next.0
  @backstage/plugin-techdocs-node                  1.3.1-next.0
  @backstage/plugin-techdocs-react                 1.0.4-next.1
  @backstage/plugin-techdocs                       1.3.2-next.1
  @backstage/plugin-user-settings                  0.4.8-next.1
  @backstage/release-manifests                     0.0.6-next.0
  @backstage/test-utils                            1.2.0-next.1
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
```
</details>

### PR DESCRIPTION
**Pull request by benjdlambert**
Fixes #13502

### CODE REVIEW

File: packages/cli/config/jest.js [hunk 1]
```diff
@@ -131,10 +131,38 @@ async function getProjectConfig(targetPath, displayName) {
...
+            },
+            transform: {
+              react: {
+                runtime: 'automatic',
+              },
+            },
+          },
+        },
+      ],
+      '\\.(t|j)sx?$': [
```
Comment (Rugvip) [inline packages/cli/config/jest.js:151]: Probably better to split this into two blocks I think? don't think it's safe to use the typescript parser for JSX

File: packages/cli/config/jest.js [hunk 2]
```diff
@@ -131,10 +131,38 @@ async function getProjectConfig(targetPath, displayName) {
...
+        require.resolve('./jestSwcTransform'),
+        {
+          sourceMaps: envOptions.enableSourceMaps || envOptions.nextTests,
+          jsc: {
+            parser: {
+              syntax: 'typescript',
+              tsx: true,
```
Comment (Rugvip) [inline packages/cli/config/jest.js:158]: realized we can't have `tsx` enabled for regular `.ts` files as well. There's syntax you can use in `.ts` but not `.tsx`, like `<string> foo`

File: packages/cli/config/jest.js [hunk 3]
```diff
@@ -131,10 +131,55 @@ async function getProjectConfig(targetPath, displayName) {
     },
 
     transform: {
-      '\\.(js|jsx|ts|tsx|mjs|cjs)$': [
+      '\\.(mjs|cjs)$': [
         require.resolve('./jestSwcTransform'),
         {
           sourceMaps: envOptions.enableSourceMaps || envOptions.nextTests,
+          jsc: {
+            target: 'es2019',
```
Comment (Rugvip) [inline packages/cli/config/jest.js:139]: should drop this one right?

File: packages/cli/config/jest.js (hunk 2 above)
Comment (benjdlambert) [reply packages/cli/config/jest.js:190]: Something like this?

File: packages/cli/config/jest.js [hunk 4]
```diff
@@ -131,10 +131,70 @@ async function getProjectConfig(targetPath, displayName) {
...
+            },
+            transform: {
+              react: {
+                runtime: 'automatic',
+              },
+            },
+          },
+        },
+      ],
+      '\\.jsx?$': [
```
Comment (Rugvip) [inline packages/cli/config/jest.js:150]: could move `.js` to the one above right?

File: packages/cli/config/jest.js [hunk 5]
```diff
@@ -131,10 +131,70 @@ async function getProjectConfig(targetPath, displayName) {
...
+          sourceMaps: envOptions.enableSourceMaps || envOptions.nextTests,
+          jsc: {
+            parser: {
+              syntax: 'typescript',
+            },
+            transform: {
+              react: {
+                runtime: 'automatic',
+              },
+            },
```
Comment (Rugvip) [inline packages/cli/config/jest.js:179]: No react in this one

File: packages/cli/config/jest.js (hunk 2 above)
Comment (Rugvip) [reply packages/cli/config/jest.js:190]: yep!

**Comment by benjdlambert** [comment]
@Rugvip maybe we merge #13503 into this branch?

**Comment by UsainBloot** [comment]
The `jsc` config for each pattern for tests is also duplicated and required by the swc loader in the bundler, extending on top of it.

What is your opinion on a base `jsc` config for each pattern which swc loader can extend from? Is it worth coupling the test and bundler transforms?

**Review by freben — APPROVED** [review]
LGTM at least!

**Comment by Rugvip** [comment]
@benjdlambert yep, I'd say go ahead and cherry-pick #13503

File: packages/cli/config/jest.js [hunk 6]
```diff
@@ -131,10 +131,65 @@ async function getProjectConfig(targetPath, displayName) {
...
           sourceMaps: envOptions.enableSourceMaps || envOptions.nextTests,
+          jsc: {
+            parser: {
+              syntax: 'ecmascript',
+            },
+            transform: {
+              react: {
+                runtime: 'automatic',
+              },
+            },
```
Comment (Rugvip) [inline packages/cli/config/jest.js:146]: Should not have any react transforms here right?

**Review by Rugvip — CHANGES_REQUESTED** [review]

File: packages/cli/config/jest.js (hunk 6 above)
Comment (Rugvip) [reply packages/cli/config/jest.js]: Fix'd this too, let's see if it builds 🤞

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**